### ISSUE TITLE
Allow virtual hosted s3 bucket regardless of URL pattern

### ISSUE THREAD

**Issue by sam6258** [ClickHouse/ClickHouse#72110]
**Company or project name**

> An s3 compatible object storage service

**Use case**

> I have an s3 compatible object storage service that does not fit the existing [regex](https://github.com/ClickHouse/ClickHouse/blob/456a41ee4249ace28e36ad65a42c320bb1fdd144/src/IO/S3/URI.cpp#L40) for virtually hosted buckets. The service does not support path style addressing as it is eventually being deprecated.

**Describe the solution you'd like**

> I would like to have a configuration option to influence the addressing style of my bucket. For example, the aws s3 CLI has `addressing_style = virtual`. The requested feature should bypass (or pass) the regex check and  use `mybucket.mydomain.com` as the authority. Note that `mydomain.com` may not require a region in a setup such as this.

**Describe alternatives you've considered**

> I can create a new bucket with a name of `mybucket.s3` but would have to migrate data to this new bucket. This also forces me to name my buckets with an `s3` postfix.
> I can match the desired regex and post-process the request before it goes to object storage, but this seems fragile and likely difficult to implement/orchestrate.

**Comment by alexey-milovidov** [comment]
I did a similar approach here: https://github.com/ClickHouse/ClickHouse/pull/68082/files#diff-943d9334c5caf8ee1cd97aaed3086c5baaa5f39d2623dd7c19e896a6935f57b1R140

If it does not match the list of known services, it will assume the bucket named `default`

**Comment by sam6258** [comment]
So it sounds like if `default` could be customized here, that would satisfy this use case, right?

**Comment by alexey-milovidov** [comment]
Yes. The simplest would be to create a global configuration value, e.g., `s3_default_bucket_name`. But maybe it is necessary to make it configurable on a per-domain level.

**Comment by sam6258** [comment]
I think `is_virtual_hosted_style = false` will be `mydomain.com/mybucket` where `is_virtual_hosted_style = true` will be `mybucket.mydomain.com` based on how the `Aws::S3::S3Client` is configured with `client_settings_.use_virtual_addressing`. If we provide a bucket override, I dont think it will play nice with the client code still if we don't set `is_virtual_hosted_style = true`.

**Comment by ADustyOldMuffin** [comment]
@alexey-milovidov For fun, is there any reason you want to detect these things inside of the URL instead of doing something like specifying a bucket and an addressing style with a base URL? That way you don't need regex parsing. I haven't dug deep into the code but I'm unsure why using regex to detect this would be desired over just building the URL.

This would be extremely common among a lot of Object Storage like APIs other than S3 itself and it seems like a pain to try and detect all of this via a regex.
